cccatalog-api/cccatalog/api/models.py

```diff
@@ -285,15 +282,54 @@ class OAuth2Verification(models.Model):
     code = models.CharField(max_length=256, db_index=True)
 
 
-class ReportImage(models.Model):
+class MatureImage(models.Model):
+    """ Stores all images that have been flagged as 'mature'. """
+    identifier = models.UUIDField(
+        unique=True,
+        primary_key=True
+    )
+    created_on = models.DateTimeField(auto_now_add=True)
+
+
+class ImageReport(models.Model):
     REPORT_CHOICES = [
         ('adult', 'adult'),
         ('dmca', 'dmca'),
         ('other', 'other')
     ]
+
+    STATUS_CHOICES = [
+        ('pending_review', 'pending_review'),
+        ('mature_filter', 'mature_filter'),
+        ('deindex', 'deindex'),
+        ('do_nothing', 'do_nothing')
+    ]
     identifier = models.UUIDField()
-    reason = models.CharField(max_length=10, choices=REPORT_CHOICES)
+    reason = models.CharField(max_length=20, choices=REPORT_CHOICES)
     description = models.TextField(max_length=500, blank=True, null=True)
+    status = models.CharField(
+        max_length=20, choices=STATUS_CHOICES, default='pending'
+    )
 
     class Meta:
         db_table = 'nsfw_reports'
+
+    def save(self, *args, **kwargs):
+        update_required = {'mature_filter', 'deindex'}
+        if self.status in update_required:
+            es = search_controller.es
+            img = Image.objects.get(identifier=self.identifier)
+            es_id = img.id
+            if self.status == 'mature_filter':
+                MatureImage(identifier=self.identifier).save()
+                es.update(
+                    index='image',
+                    id=es_id,
+                    body={'doc': {'mature': True}}
+                )
+            elif self.status == 'deindex':
+                img.delete()
+                DeletedImage(identifier=self.identifier).save()
+                es.delete(index='image', id=es_id)
+            es.indices.refresh(index='image')
```

Comment:
I've taken your style advice and extracted the statuses asg constants, I think it will help prevent subtle bugs. I would prefer to use an enum but calling `(ReportStatus.MATURE_FILTERED.value, ReportStatus.MATURE_FILTERED.value)` in STATUS_CHOICES is too clunky.

REPORT_CHOICES aren't used for any internal logic at the moment, they're more markers for reviewers, so I'm going to leave them hardcoded for now.

>Somebody complains about some content which happens to be 'mature' being non-CC-licensed. So, we deindex it. But, we get a subsequent complaint from the creator that the content is CC-licensed, and they're upset it was removed, so we un-deindex it. But at that point, it's not marked mature, which it should be.
>It might make sense (perhaps in a future PR) to have some sort of checkbox UI instead of he drop-down to facilitate toggling on/off different combinations of flags.

An ImageReport represents a single complaint by a user; it is possible for multiple reports to be filed against a single image. When a status has been assigned to a report, it is not possible to revoke it through the already processed report (it doesn't make sense to go from `mature_filtered` back to `pending_review`; the item has been reviewed already). This is why it is not possible to change the status of a report after it has been processed.

To remove the mature flag from a work, someone can use the admin UI to remove the image from the `MatureImage` table, and `mature = False` will automatically be set in Elasticsearch (see `MatureImage.delete`). We can do the same for deindexed work in the future if the requirement arises; deleting something from the `DeletedImages` table will result in the work being restored next time we reindex our data.